File: ReactAndroid/src/main/java/com/facebook/react/Experiments.java
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.facebook.react;

import java.util.HashSet;
import java.util.Set;


/**
 * This class exposes a way to enable/disable experimental features in React Native.
 */
public class Experiments {

  // Enables Hot Module Replacement menu item in the developer menu
  public static String HOT_MODULE_REPLACEMENT = "HOT_MODULE_REPLACEMENT";

  private static Set<String> enabledExperiments;

  /**
   * Enables an experiment
   */
  public static void enableExperiment(String experiment) {
    if (enabledExperiments == null) {
      enabledExperiments = new HashSet<>();
    }
    enabledExperiments.add(experiment);
  }

  /**
   * Disables an experiment
   */
  public static void disableExperiment(String experiment) {
    if (enabledExperiments != null) {
      enabledExperiments.remove(experiment);
    }
  }

  /**
   * Checks if an experiment is enabled
   */
  public static boolean isExperimentEnabled(String experiment) {
    return enabledExperiments != null && enabledExperiments.contains(experiment);
  }
}

File: packages/react-native/ReactCommon/yoga/yoga/internal/experiments.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "experiments.h"
#include "experiments-inl.h"

namespace facebook {
namespace yoga {
namespace internal {

namespace detail {

std::bitset<sizeof(int)> enabledExperiments = 0;

} // namespace detail

void enable(Experiment experiment) {
  detail::enabledExperiments.set(static_cast<size_t>(experiment));
}

void disable(Experiment experiment) {
  detail::enabledExperiments.reset(static_cast<size_t>(experiment));
}

bool toggle(Experiment experiment) {
  auto bit = static_cast<size_t>(experiment);
  auto previousState = detail::enabledExperiments.test(bit);
  detail::enabledExperiments.flip(bit);
  return previousState;
}

} // namespace internal
} // namespace yoga
} // namespace facebook
